File: app/src/main/java/com/hiddenramblings/tagmo/amiibo/tagdata/Checksum.java
/*
 * ====================================================================
 * SSBU_Amiibo Copyright (c) 2021 odwdinc
 * src/ssbu_amiibo/amiibo_class.py
 * smash-amiibo-editor Copyright (c) 2021 jozz024
 * utils/ssbu_amiibo.py
 * Copyright (C) 2022 AbandonedCart @ TagMo
 * ====================================================================
 */

package com.hiddenramblings.tagmo.amiibo.tagdata;

public class Checksum {

    private final byte[] u0;

    public Checksum() {
        int p0 = 0xEDB88320 | 0x80000000;

        u0 = new byte[0x100];
        int i = 0x1;
        while ((i & 0xFF) != 0) {
            int t0 = i;
            for (int x = 0; x < 0x8; x++) {
                boolean b = (t0 & 0x1) != 0;
                t0 >>= 0x1;
                if (b) t0 ^= p0;
            }
            u0[i] = (byte) (t0);
            i += 0x1;
        }
    }

    public int generate(byte[] appData) {
        byte[] checksum = new byte[0xD4];
        System.arraycopy(appData, 0x04, checksum, 0, checksum.length);

        int t = 0x0;
        for (byte k : checksum) {
            t = ((t >> 0x8) ^ u0[(k ^ t) & 0xFF]);
        }
        return t ^ 0xFFFFFFFF;
    }
}

File: app/src/main/java/com/hiddenramblings/tagmo/amiibo/tagdata/Checksum.kt
/*
 * ====================================================================
 * SSBU_Amiibo Copyright (c) 2021 odwdinc
 * src/ssbu_amiibo/amiibo_class.py
 * smash-amiibo-editor Copyright (c) 2021 jozz024
 * utils/ssbu_amiibo.py
 * Copyright (C) 2022 AbandonedCart @ TagMo
 * ====================================================================
 */
package com.hiddenramblings.tagmo.amiibo.tagdata

import java.nio.ByteBuffer

class Checksum {
    private val u0: ByteArray

    init {
        val p0 = -0x12477ce0 or -0x80000000
        u0 = ByteArray(0x100)
        var i = 0x1
        while (i and 0xFF != 0) {
            var t0 = i
            for (x in 0..0x7) {
                val b = t0 and 0x1 != 0
                t0 = t0 shr 0x1
                if (b) t0 = t0 xor p0
            }
            u0[i] = t0.toByte()
            i += 0x1
        }
    }

    fun generate(appData: ByteBuffer): Int {
        val checksum = ByteArray(212) // 0xD4
        System.arraycopy(appData.array(), 0xE0, checksum, 0, checksum.size)
        var t = 0x0
        checksum.forEach {
            t = t shr 0x8 xor u0[it.toInt() xor t and 0xFF].toInt()
        }
        return t xor -0x1
    }
}